client/src/storage/device.rs

```diff
@@ -247,14 +673,336 @@ impl SenderKeyStore for DeviceSenderKeyStore {
         distribution_id: Uuid,
         record: &SenderKeyRecord,
     ) -> Result<(), SignalProtocolError> {
-        todo!()
+        let addr = format!("{}:{}", sender, distribution_id);
+        let rec = BASE64_STANDARD.encode(record.serialize()?);
+
+        sqlx::query!(
+            r#"
+            INSERT INTO DeviceSenderKeyStore (address, sender_key_record)
+            VALUES (?, ?)
+            ON CONFLICT(address) DO UPDATE SET sender_key_record = ?
+            "#,
+            addr,
+            rec,
+            rec
+        )
+        .execute(&self.pool)
+        .await
+        .map(|_| ())
+        .map_err(|err| SignalProtocolError::InvalidArgument(format!("{}", err)))
     }
 
     async fn load_sender_key(
         &mut self,
         sender: &ProtocolAddress,
         distribution_id: Uuid,
     ) -> Result<Option<SenderKeyRecord>, SignalProtocolError> {
-        todo!()
+        let addr = format!("{}:{}", sender, distribution_id);
+
+        match sqlx::query!(
+            r#"
+            SELECT
+                sender_key_record
+            FROM
+                DeviceSenderKeyStore
+            WHERE
+                address = ?
+            "#,
+            addr
+        )
+        .fetch_one(&self.pool)
+        .await
+        {
+            Ok(row) => SenderKeyRecord::deserialize(
+                BASE64_STANDARD
+                    .decode(row.sender_key_record)
+                    .map_err(|err| SignalProtocolError::InvalidArgument(format!("{}", err)))?
+                    .as_slice(),
+            )
+            .map(|res| Some(res)),
+            Err(_) => Ok(None),
+        }
+    }
+}
+
+#[cfg(test)]
+mod device_protocol_test {
+    use common::utils::time_now;
+    use libsignal_core::{Aci, ProtocolAddress};
+    use libsignal_protocol::{
+        kem::KeyType, Direction, GenericSignedPreKey, IdentityKeyPair, IdentityKeyStore, KeyPair,
+        KyberPreKeyId, KyberPreKeyRecord, KyberPreKeyStore, PreKeyId, PreKeyRecord, PreKeyStore,
+        SessionRecord, SessionStore, SignedPreKeyId, SignedPreKeyRecord, SignedPreKeyStore,
+    };
+    use rand::rngs::OsRng;
+    use sqlx::{sqlite::SqlitePoolOptions, SqlitePool};
+    use uuid::Uuid;
+
+    use crate::storage::device::{
+        DeviceIdentityKeyStore, DeviceKyberPreKeyStore, DevicePreKeyStore, DeviceSessionStore,
+        DeviceSignedPreKeyStore,
+    };
+
+    async fn connect() -> SqlitePool {
+        dotenv::dotenv().ok();
+        let db_url =
+            std::env::var("DATABASE_URL").expect("Expected to read database url from .env file");
```

Comment:
Why not use the test database?

Reply:
Når vi kører test mes SQLite, så bør vi bare sætter DATABASE_URL til :memory: